src/i965_encoder_vp8.c

```diff
@@ -0,0 +1,6244 @@
+            vp8_context->down_scaled_width_in_mb4x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+
+        if (vp8_context->down_scaled_height_4x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_height_in_mb4x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_height_4x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_height_in_mb4x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+    } else if (vp8_context->down_scaled_width_16x < vp8_context->min_scaled_dimension ||
+               vp8_context->down_scaled_width_in_mb16x < vp8_context->min_scaled_dimension_in_mbs ||
+               vp8_context->down_scaled_height_16x < vp8_context->min_scaled_dimension ||
+               vp8_context->down_scaled_height_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+        if (vp8_context->down_scaled_width_16x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_width_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_width_16x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_width_in_mb16x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+
+        if (vp8_context->down_scaled_height_16x < vp8_context->min_scaled_dimension ||
+            vp8_context->down_scaled_height_in_mb16x < vp8_context->min_scaled_dimension_in_mbs) {
+
+            vp8_context->down_scaled_height_16x = vp8_context->min_scaled_dimension;
+            vp8_context->down_scaled_height_in_mb16x = vp8_context->min_scaled_dimension_in_mbs;
+        }
+    }
+}
+
+static void
+i965_encoder_vp8_free_surfaces(void **data)
+{
+    struct i965_encoder_vp8_surface *vp8_surface;
+
+    if (!data || *data)
+        return;
+
+    vp8_surface = *data;
+
+    if (vp8_surface->scaled_4x_surface_obj) {
+        i965_DestroySurfaces(vp8_surface->ctx, &vp8_surface->scaled_4x_surface_id, 1);
+        vp8_surface->scaled_4x_surface_id = VA_INVALID_SURFACE;
+        vp8_surface->scaled_4x_surface_obj = NULL;
+    }
+
+    if (vp8_surface->scaled_16x_surface_obj) {
+        i965_DestroySurfaces(vp8_surface->ctx, &vp8_surface->scaled_16x_surface_id, 1);
+        vp8_surface->scaled_16x_surface_id = VA_INVALID_SURFACE;
+        vp8_surface->scaled_16x_surface_obj = NULL;
+    }
+}
+
+static void
+i965_encoder_vp8_allocate_surfaces(VADriverContextP ctx,
+                                   struct intel_encoder_context *encoder_context,
+                                   struct object_surface *obj_surface)
+{
+    struct i965_driver_data *i965 = i965_driver_data(ctx);
+    struct i965_encoder_vp8_context *vp8_context = encoder_context->vme_context;
+    struct i965_encoder_vp8_surface *vp8_surface;
+    int down_scaled_width_4x, down_scaled_height_4x;
+    int down_scaled_width_16x, down_scaled_height_16x;
+
+    if (!obj_surface || obj_surface->private_data)
```

Comment:
I don't think so. If there is no relationship between two different contexts, user needn't to reuse a surface between the two contexts. so the normal usage is a surface shared by a decoder & a encoder, not two encoders:
for example:
a> the first surface (I frame) is output from a H.264 decoding context.
b> this frame is used as an input of a VP8 encoding context.
c> this surface is used as a reference in a H.264 decoding.
free&reallocation will still break H.264 decoding and VP8 encoding.